code/controllers/subsystem/dbcore.dm

```diff
@@ -0,0 +1,246 @@
+SUBSYSTEM_DEF(dbcore)
+	name = "Database"
+	flags = SS_NO_INIT|SS_NO_FIRE
+	init_order = INIT_ORDER_DBCORE
+	var/const/FAILED_DB_CONNECTION_CUTOFF = 5
+
+	var/const/Default_Cursor = 0
+	var/const/Client_Cursor = 1
+	var/const/Server_Cursor = 2
+	//conversions
+	var/const/TEXT_CONV = 1
+	var/const/RSC_FILE_CONV = 2
+	var/const/NUMBER_CONV = 3
+	//column flag values:
+	var/const/IS_NUMERIC = 1
+	var/const/IS_BINARY = 2
+	var/const/IS_NOT_NULL = 4
+	var/const/IS_PRIMARY_KEY = 8
+	var/const/IS_UNSIGNED = 16
+// TODO: Investigate more recent type additions and see if I can handle them. - Nadrew
+
+	var/_db_con// This variable contains a reference to the actual database connection.
+	var/failed_connections = 0
+
+/datum/controller/subsystem/dbcore/PreInit()
+	_db_con = _dm_db_new_con()
+
+/datum/controller/subsystem/dbcore/Recover()
+	_db_con = SSdbcore._db_con
+
+/datum/controller/subsystem/dbcore/Shutdown()
+	// This is as clsoe as we can get to the true round end before Disconnect() without changing where it's called, deafeating the reason this is a subsystem
+	if(SSdbcore.Connect())
+		var/sql_station_name = sanitizeSQL(station_name())
+		var/datum/DBQuery/query_round_end = SSdbcore.NewQuery("INSERT INTO [format_table_name("round")] (shutdown_datetime, game_mode_result, end_state, station_name) VALUES (Now(), '[ticker.mode_result]', '[ticker.end_state]', '[sql_station_name]') WHERE id = [GLOB.round_id]")
+		query_round_end.Execute()
+	if(IsConnected())
+		Disconnect()
+
+//nu
+/datum/controller/subsystem/dbcore/can_vv_get(var_name)
+	return var_name != "_db_con" && ..()
+
+/datum/controller/subsystem/dbcore/vv_edit_var(var_name, var_value)
+	if(var_name == "_db_con")
+		return FALSE
+	return ..()
+
+/datum/controller/subsystem/dbcore/proc/Connect()
+	if(IsConnected())
+		return TRUE
+
+	if(failed_connections > FAILED_DB_CONNECTION_CUTOFF)	//If it failed to establish a connection more than 5 times in a row, don't bother attempting to connect anymore.
+		return FALSE
+	
+	if(!config.sql_enabled)
+		return FALSE
+
+	var/user = global.sqlfdbklogin
+	var/pass = global.sqlfdbkpass
+	var/db = global.sqlfdbkdb
+	var/address = global.sqladdress
+	var/port = global.sqlport
+
+	doConnect("dbi:mysql:[db]:[address]:[port]", user, pass)
+	. = IsConnected()
+	if(!.)
+		log_sql("Connect() failed | [ErrorMsg()]")
+		++failed_connections
+
+/datum/controller/subsystem/dbcore/proc/SetRoundID()
+	if(!Connect())
+		return
+	var/datum/DBQuery/query_round_initialize = SSdbcore.NewQuery("INSERT INTO [format_table_name("round")] (initialize_datetime, server_ip, server_port) VALUES (Now(), INET_ATON(IF('[world.internet_address]' LIKE '', '0', '[world.internet_address]')), '[world.port]')")
+	query_round_initialize.Execute()
+	var/datum/DBQuery/query_round_last_id = SSdbcore.NewQuery("SELECT LAST_INSERT_ID()")
+	query_round_last_id.Execute()
+	if(query_round_last_id.NextRow())
+		GLOB.round_id = query_round_last_id.item[1]
```

Comment:
Shouldn’t this be a query + 1, or it’s gonna use the same round ID as the previous round. And if it’s empty does it not pull anything? This was problematic when I tried to do it so heads up

Reply:
if(query_round_last_id.NextRow()) IDEALLY should be the stop here is theres nothing there. and from what i can see from the following procs its updating the id from start and end round. also by the vars here "query_round_last_id" that is SPECIFCIALLY what its doing.setting it to the last id.

Reply:
This works because of the INSERT statement above it. There's no table context to this SQL SELECT fragment, so it'll give the last auto-increment number overall, which ideally was whatever was assigned to the INSERT statement above. This is why, in tgstation's dbcore.dm, this is run with async = false-- it relies on synchronous execution of these statements, so that the round_id actually matches the round that ends up in the `round` table.